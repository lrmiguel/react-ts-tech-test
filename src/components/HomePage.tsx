import { PaginatedEstablishmentsTable } from "./PaginatedEstablishmentsTable";
import Background from "../static/logo.svg";
import { AuthoritiesDropDown } from "./AuthoritiesDropDown";
import { QueryClient, QueryClientProvider } from 'react-query';
import { useState } from "react";

const queryClient = new QueryClient();

const logoStyle: { [key: string]: string | number } = {
  width: "640px",
  height: "200px",
  background: `transparent url(${Background}) no-repeat center`,
  margin: "20px auto",
};
const HomePage = () => {
  const [selectedAuthority, setSelectedAuthority] = useState<string | null>(null);
  
  return (
    <div>
      <header style={logoStyle} />
      <QueryClientProvider client={queryClient}>
        <AuthoritiesDropDown onChange={setSelectedAuthority} />
        <PaginatedEstablishmentsTable authority={selectedAuthority ?? undefined} />
      </QueryClientProvider>
    </div>
  );
};

export default HomePage;
