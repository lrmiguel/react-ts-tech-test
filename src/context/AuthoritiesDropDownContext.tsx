import { createContext, useContext } from "react";

interface AuthoritiesDropdownContextProps {
    menuRef: React.MutableRefObject<HTMLDivElement | null>;
    hasMore: boolean | void;
    handleLoadMore: () => void;
    isFetchingNextPage: boolean;
}

export const AuthoritiesDropdownContext = createContext<AuthoritiesDropdownContextProps | undefined>(undefined);

export const useAuthoritiesDropdown = () => {
    const ctx = useContext(AuthoritiesDropdownContext);
    if (!ctx) throw new Error("useAuthoritiesDropdown must be used within AuthoritiesDropdownContext.Provider");
    return ctx;
  };