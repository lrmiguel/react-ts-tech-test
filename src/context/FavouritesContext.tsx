import React, { createContext, useContext, useState } from "react";

export interface Favourite {
  id: string;
  name: string;
  rating: string;
}

interface FavouritesContextType {
  favourites: Favourite[];
  toggleFavourite: (fav: Favourite) => void;
  removeFavourite: (id: string) => void;
  isFavourited: (id: string) => boolean;
}

const FavouritesContext = createContext<FavouritesContextType | undefined>(undefined);

export const FavouritesProvider: React.FC<{ children: React.ReactNode }> = ({ children }) => {
  const [favourites, setFavourites] = useState<Favourite[]>([]);

  const toggleFavourite = (fav: Favourite) => {
    setFavourites((prev) =>
      prev.some((f) => f.id === fav.id)
        ? prev.filter((f) => f.id !== fav.id)
        : [...prev, fav]
    );
  };

  const removeFavourite = (id: string) => {
    setFavourites((prev) => prev.filter((f) => f.id !== id));
  };

  const isFavourited = (id: string) => {
    return favourites.some((f) => f.id === id);
  };
  return (
    <FavouritesContext.Provider value={{ favourites, toggleFavourite, removeFavourite, isFavourited }}>
      {children}
    </FavouritesContext.Provider>
  );
};

export const useFavourites = () => {
  const ctx = useContext(FavouritesContext);
  if (!ctx) throw new Error("useFavourites must be used within FavouritesProvider");
  return ctx;
};
