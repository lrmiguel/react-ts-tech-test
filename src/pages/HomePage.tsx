import { PaginatedEstablishmentsTable } from "../components/establishments/PaginatedEstablishmentsTable";
import { useState } from "react";
import PageLayout from "./PageLayout";
import { AuthoritiesDropDown } from "../components/authorities/AuthoritiesDropDown";

const HomePage = () => {
  const [selectedAuthority, setSelectedAuthority] = useState<string | null>(null);
  
  return (
    <PageLayout>
      <AuthoritiesDropDown onChange={setSelectedAuthority} />
      <PaginatedEstablishmentsTable authority={selectedAuthority ?? undefined} />
    </PageLayout>
  );
};

export default HomePage;
