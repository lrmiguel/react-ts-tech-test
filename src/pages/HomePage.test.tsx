import { render } from "@testing-library/react";
import { QueryClient, QueryClientProvider } from "react-query";
import HomePage from "./HomePage";
import { MemoryRouter } from "react-router-dom";
import { FavouritesProvider } from "../context/FavouritesContext";

const createQueryClient = () => {
  return new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
        cacheTime: 0,
        staleTime: 0,
      },
    },
  });
};

const renderWithProviders = () => {
  return render(
    <MemoryRouter>
      <QueryClientProvider client={createQueryClient()}>
        <FavouritesProvider>
          <HomePage />
        </FavouritesProvider>
      </QueryClientProvider>
    </MemoryRouter>
  );
};

describe("HomePage", () => {
  it("matches snapshot", () => {
    const { container } = renderWithProviders();
    expect(container).toMatchSnapshot();
  });
});