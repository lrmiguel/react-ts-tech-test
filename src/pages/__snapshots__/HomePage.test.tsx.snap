// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage matches snapshot 1`] = `
<div>
  <div
    style="align-items: center; color: white; display: flex; flex-direction: column; height: 100vh;"
  >
    <header
      style="width: 640px; height: 200px; background: transparent url(logo.svg) no-repeat center; background-size: contain; flex-shrink: 0; margin-bottom: 20px;"
    />
    <div
      style="color: white; font-size: 20px; width: calc(50rem + 40px);"
    >
      <label
        for="authoritiesDropDown"
        style="font-weight: bold;"
      >
        Authorities:
      </label>
      <div
        class="css-b62m3t-container"
        id="authoritiesDropDown"
      >
        <span
          class="css-1f43avz-a11yText-A11yText"
          id="react-select-2-live-region"
        />
        <span
          aria-atomic="false"
          aria-live="polite"
          aria-relevant="additions text"
          class="css-1f43avz-a11yText-A11yText"
          role="log"
        />
        <div
          class="css-1asr2dn-control"
        >
          <div
            class="css-hlgwow"
          >
            <div
              class="css-1550uzu-placeholder"
              id="react-select-2-placeholder"
            >
              Search or select...
            </div>
            <div
              class="css-19bb58m"
              data-value=""
            >
              <input
                aria-activedescendant=""
                aria-autocomplete="list"
                aria-describedby="react-select-2-placeholder"
                aria-expanded="false"
                aria-haspopup="true"
                autocapitalize="none"
                autocomplete="off"
                autocorrect="off"
                class=""
                id="react-select-2-input"
                role="combobox"
                spellcheck="false"
                style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                tabindex="0"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="css-1wy0on6"
          >
            <div
              aria-hidden="true"
              class="css-pogzpp-loadingIndicator"
            >
              <span
                class="css-1icxkl9-LoadingDot"
              />
              <span
                class="css-1oazjx7-LoadingDot"
              />
              <span
                class="css-1wonz24-LoadingDot"
              />
            </div>
            <span
              class="css-1u9des2-indicatorSeparator"
            />
            <div
              aria-hidden="true"
              class="css-1xc3v61-indicatorContainer"
            >
              <svg
                aria-hidden="true"
                class="css-tj5bde-Svg"
                focusable="false"
                height="20"
                viewBox="0 0 20 20"
                width="20"
              >
                <path
                  d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      style="background: rgb(130, 199, 175); color: white; padding: 10px 20px; width: 50rem; height: 29rem;"
    >
      <h2>
        Food Hygiene Ratings
      </h2>
      <table>
        <tbody>
          <tr>
            <th
              style="font-size: 20px; text-align: left; padding-bottom: 10px; padding-right: 30px;"
            >
              Fav
            </th>
            <th
              style="font-size: 20px; text-align: left; padding-bottom: 10px; width: 37.5rem;"
            >
              Business Name
            </th>
            <th
              style="font-size: 20px; text-align: left; padding-bottom: 10px;"
            >
              Rating Value
            </th>
          </tr>
          <tr
            style="font-size: 20px; height: 19.15rem;"
          >
            <td
              colspan="3"
              style="font-weight: bold; text-align: center;"
            >
              Loading...
            </td>
          </tr>
        </tbody>
      </table>
      <nav>
        <button
          disabled=""
          style="margin: 10px 5px;"
          type="button"
        >
          -
        </button>
        1
        <button
          disabled=""
          style="margin: 10px 5px;"
          type="button"
        >
          +
        </button>
      </nav>
    </div>
  </div>
</div>
`;
